internal/controlplane/handlers_authz.go

```diff
@@ -216,9 +218,100 @@ func (s *Server) ListRoleAssignments(
 		return nil, status.Errorf(codes.Internal, "error getting role assignments: %v", err)
 	}
 
+	as, err = s.mergeMatchedRoleMappingstoAssignments(ctx, projectID, as)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error matching role assignments to mappings: %v", err)
+	}
+
+	unm, err := s.getUnmatchedMappings(ctx, projectID)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error getting unmatched mappings: %v", err)
+	}
+
 	return &minder.ListRoleAssignmentsResponse{
-		RoleAssignments: as,
+		RoleAssignments:   as,
+		UnmatchedMappings: unm,
 	}, nil
+
+}
+
+func (s *Server) mergeMatchedRoleMappingstoAssignments(
+	ctx context.Context,
+	projectID uuid.UUID,
+	assignmentsToMerge []*minder.RoleAssignment,
+) ([]*minder.RoleAssignment, error) {
+	res, err := s.store.ListResolvedMappedRoleGrantsForProject(ctx, projectID)
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		return nil, err
+	} else if err != nil && errors.Is(err, sql.ErrNoRows) {
+		return assignmentsToMerge, nil
+	}
+
+	for idx := range assignmentsToMerge {
+		aToM := assignmentsToMerge[idx]
+
+		foundIdx := -1
+
+		for idx, r := range res {
+			if aToM.Role == r.Role && aToM.Subject == r.ResolvedSubject.String {
+				mID := r.ID.String()
+				ctom := &structpb.Struct{}
+
+				if err := json.Unmarshal(r.ClaimMappings, ctom); err != nil {
+					zerolog.Ctx(ctx).Err(err).Str("mapping-id", r.ID.String()).
+						Msg("couldn't marshal mappings into JSON")
+					continue
+				}
+
+				aToM.Mapping = &minder.RoleAssignment_Mapping{
+					Id:            &mID,
+					ClaimsToMatch: ctom,
+				}
+
+				foundIdx = idx
+				break
+			}
+		}
+
+		if foundIdx >= 0 {
+			res = append(res[:foundIdx], res[foundIdx+1:]...)
+		}
+	}
+
+	return assignmentsToMerge, nil
```

Comment:
`assignmentsToMerge` gets decorated with the information from the role mappings, we'd be able to see what mapping did a specific assignment come from. So it's all done inside of that array. `res` gets iterated to determine if it matches a specific mapping, that array gets trimmed to avoid further iterations on things that were solved or merged already.